src/main/java/javax/jmdns/impl/ServiceInfoImpl.java

```diff
@@ -764,15 +764,14 @@ public void updateRecord(final DNSCache dnsCache, final long now, final DNSEntry
                     ServiceEvent event = new ServiceEventImpl(dns, this.getType(), this.getName(), this.clone());
                     dns.handleServiceResolved(event);
                 }
+                // This is done, to notify the wait loop in method JmDNS.waitForInfoData(ServiceInfo info, int timeout);
+                synchronized (this) {
+                    this.notifyAll();
+                }
             } else {
                 logger.debug("JmDNS not available.");
             }
         }
-
-        // This is done, to notify the wait loop in method JmDNS.waitForInfoData(ServiceInfo info, int timeout);
-        synchronized (this) {
-            this.notifyAll();
-        }
```

Comment:
I think if change the notifyAll() here may cause problem:
If notifyAll() when service info changed but the service info has not changed any more, it may be no chance to wait up the waiters (without wait timeout) of service info then. 

I think It's better to change the JmDNSImpl.waitForInfoData to wait enough time instead of changing the notifyAll() here.

Reply:
I checked with @zhangl22000, the service info needs to call notifyAll when the dnsEntry is related this service info.

Reply:
Both fixes are fine. Fix option 1: when service info has changes, notifyAll() is called.  As waiForInfoData() only wait when service info has no data). Fix option 2: notifyAll() when this service info is updated  (updated, but may not change).